### PR TITLE
KAFKA-16197: Print Connect worker specific logs on poll timeout expiry

### PR DESCRIPTION
**Pull request by vamossagar12**
Please refer to the description in [KAFKA-16197](https://issues.apache.org/jira/browse/KAFKA-16197). 

Note that there are 2 more instances [here](https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java#L1131) and [here](https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java#L1079) which print consumer specific changes but that had changes leaking into ConsumerCoordinator and other tests for it because of which I didn't include it as part of this PR.

### CODE REVIEW

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/WorkerCoordinatorTest.java
Comment (vamossagar12) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/WorkerCoordinatorTest.java]: I increased `sessionTimeoutMs` and `heartbeatIntervalMs` by small numbers so as to be able to mock a poll timeout expiry without a lot of session timeouts and losing connectivity with Group Coordinator.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/WorkerCoordinatorTest.java
Comment (vamossagar12) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/WorkerCoordinatorTest.java]: Note that despite multiple permutations of the timeout configs, I still see one session timeout which seems hard to eliminate.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/WorkerCoordinator.java
Comment (vamossagar12) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/WorkerCoordinator.java]: The bit about `then it can be addressed by either adding more workers`, is something which has been experienced in certain cases when there are too many tasks to start which leads to poll timeout. I can remove it if it's not deemed necessary here.

**Comment by vamossagar12** [comment]
@showuon would you have time to review this? It's a smallish change. Thanks!

**Comment by showuon** [comment]
Will check it this week.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/WorkerCoordinatorTest.java
Comment (showuon) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/WorkerCoordinatorTest.java]: 1. You didn't provide HeartBeatResponse, so it'll have session timeout.
2. The `heartbeatIntervalMs` is the minimum time interval the heartbeat should wait, the real timeout for heartBeat should be sessionTimeout, so we can set `sessionTimeoutMs - 1` to make the time faster to reach `rebalanceTimeoutMs`.
3. The last poll doesn't make any sense because the poll timeout should be triggered already. Why do we need it?

What I would write is something like this, FYR:
```
public void testPollTimeoutExpiry() throws InterruptedException {

        when(configStorage.snapshot()).thenReturn(configState1);

        client.prepareResponse(FindCoordinatorResponse.prepareResponse(Errors.NONE, groupId, node));
        coordinator.ensureCoordinatorReady(time.timer(Long.MAX_VALUE));

        client.prepareResponse(joinGroupFollowerResponse(1, "member", "leader", Errors.NONE));
        client.prepareResponse(syncGroupResponse(ConnectProtocol.Assignment.NO_ERROR, "leader", configState1.offset(), Collections.emptyList(),
            Collections.singletonList(taskId1x0), Errors.NONE));

        // prepare 3 heartBeatResponses because we will trigger 3 heartBeat requests until rebalanceTimeout,
        // that is (sessionTimeoutMs - 1) * 3 > rebalanceTimeoutMs
        client.prepareResponse(new HeartbeatResponse(new HeartbeatResponseData()));
        client.prepareResponse(new HeartbeatResponse(new HeartbeatResponseData()));
        client.prepareResponse(new HeartbeatResponse(new HeartbeatResponseData()));

        try (LogCaptureAppender logCaptureAppender = LogCaptureAppender.createAndRegister(WorkerCoordinator.class)) {
            coordinator.ensureActiveGroup();
            coordinator.poll(0, () -> {
                return null;
            });

            // We keep the heartbeat thread running behind the scenes and poll frequently so that eventually
            // the time goes past now + rebalanceTimeoutMs which triggers poll timeout expiry.
            TestUtils.waitForCondition(() -> {
                // sleep until sessionTimeoutMs to trigger a heartBeat request to avoid session timeout.
                // Not sure if this will be flaky in CI because the heartbeat thread might not send out the heartBeat request in time.
                time.sleep(sessionTimeoutMs - 1);
                return logCaptureAppender.getEvents().stream().anyMatch(e -> e.getLevel().equals("WARN")) &&
                        logCaptureAppender.getEvents().stream().anyMatch(e -> e.getMessage().startsWith("worker poll timeout has expired"));
            }, "Coordinator did not poll for rebalance.timeout.ms");
        }
    }
```

Comment (vamossagar12) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/WorkerCoordinatorTest.java]: Thanks @showuon for the suggestions. I updated the test now.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/WorkerCoordinatorTest.java
Comment (vamossagar12) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/WorkerCoordinatorTest.java]: If I don't have these 2 lines, the tests become flaky. With the 2 lines added, I ran all the tests in WorkerCoordinatorTest 30 times and all the tests passes. This is needed because sometimes in the test the connectivity with coordinator goes away due to session timeout and a classcast exception gets thrown. Adding logs for referecene:

```
[2024-04-10 16:35:06,023] INFO Cluster ID: kafka-cluster (org.apache.kafka.clients.Metadata:349)
[2024-04-10 16:35:06,028] DEBUG Sending FindCoordinator request to broker localhost:1969 (id: 0 rack: null) (org.apache.kafka.connect.runtime.distributed.WorkerCoordinator:904)
[2024-04-10 16:35:06,029] DEBUG Received FindCoordinator response ClientResponse(receivedTimeMs=1712747106023, latencyMs=0, disconnected=false, timedOut=false, requestHeader=RequestHeader(apiKey=FIND_COORDINATOR, apiVersion=4, clientId=mockClientId, correlationId=0, headerVersion=2), responseBody=FindCoordinatorResponseData(throttleTimeMs=0, errorCode=0, errorMessage='', nodeId=0, host='', port=0, coordinators=[Coordinator(key='test-group', nodeId=0, host='localhost', port=1969, errorCode=0, errorMessage='NONE')])) (org.apache.kafka.connect.runtime.distributed.WorkerCoordinator:917)
[2024-04-10 16:35:06,029] INFO Discovered group coordinator localhost:1969 (id: 2147483647 rack: null) (org.apache.kafka.connect.runtime.distributed.WorkerCoordinator:936)
[2024-04-10 16:35:06,030] INFO Rebalance started (org.apache.kafka.connect.runtime.distributed.WorkerCoordinator:242)
[2024-04-10 16:35:06,030] DEBUG Heartbeat thread started (org.apache.kafka.connect.runtime.distributed.WorkerCoordinator:1481)
[2024-04-10 16:35:06,030] DEBUG Cooperative rebalance triggered. Keeping assignment null until it's explicitly revoked. (org.apache.kafka.connect.runtime.distributed.WorkerCoordinator:250)
[2024-04-10 16:35:06,030] INFO (Re-)joining group (org.apache.kafka.connect.runtime.distributed.WorkerCoordinator:604)
[2024-04-10 16:35:06,031] DEBUG Sending JoinGroup (JoinGroupRequestData(groupId='test-group', sessionTimeoutMs=30, rebalanceTimeoutMs=60, memberId='', groupInstanceId=null, protocolType='connect', protocols=[JoinGroupRequestProtocol(name='compatible', metadata=[0, 1, 0, 14, 108, 101, 97, 100, 101, 114, 85, 114, 108, 58, 56, 48, 56, 51, 0, 0, 0, 0, 0, 0, 0, 4, -1, -1, -1, -1]), JoinGroupRequestProtocol(name='default', metadata=[0, 0, 0, 14, 108, 101, 97, 100, 101, 114, 85, 114, 108, 58, 56, 48, 56, 51, 0, 0, 0, 0, 0, 0, 0, 4])], reason='')) to coordinator localhost:1969 (id: 2147483647 rack: null) (org.apache.kafka.connect.runtime.distributed.WorkerCoordinator:617)
[2024-04-10 16:35:06,031] DEBUG Received successful JoinGroup response: JoinGroupResponseData(throttleTimeMs=0, errorCode=0, generationId=1, protocolType=null, protocolName='default', leader='leader', skipAssignment=false, memberId='member', members=[]) (org.apache.kafka.connect.runtime.distributed.WorkerCoordinator:645)
[2024-04-10 16:35:06,031] DEBUG Enabling heartbeat thread (org.apache.kafka.connect.runtime.distributed.WorkerCoordinator:1449)
[2024-04-10 16:35:06,031] INFO Successfully joined group with generation Generation{generationId=1, memberId='member', protocol='default'} (org.apache.kafka.connect.runtime.distributed.WorkerCoordinator:665)
[2024-04-10 16:35:06,031] DEBUG Sending follower SyncGroup to coordinator localhost:1969 (id: 2147483647 rack: null): SyncGroupRequestData(groupId='test-group', generationId=1, memberId='member', groupInstanceId=null, protocolType='connect', protocolName='default', assignments=[]) (org.apache.kafka.connect.runtime.distributed.WorkerCoordinator:759)
[2024-04-10 16:35:06,032] DEBUG Received successful SyncGroup response: SyncGroupResponseData(throttleTimeMs=0, errorCode=0, protocolType=null, protocolName=null, assignment=[0, 0, 0, 0, 0, 6, 108, 101, 97, 100, 101, 114, 0, 14, 108, 101, 97, 100, 101, 114, 85, 114, 108, 58, 56, 48, 56, 51, 0, 0, 0, 0, 0, 0, 0, 4, 0, 0, 0, 1, 0, 10, 99, 111, 110, 110, 101, 99, 116, 111, 114, 49, 0, 0, 0, 1, 0, 0, 0, 0]) (org.apache.kafka.connect.runtime.distributed.WorkerCoordinator:826)
[2024-04-10 16:35:06,032] INFO Successfully synced group in generation Generation{generationId=1, memberId='member', protocol='default'} (org.apache.kafka.connect.runtime.distributed.WorkerCoordinator:842)
[2024-04-10 16:35:06,032] DEBUG Deserialized new assignment: Assignment{error=0, leader='leader', leaderUrl='leaderUrl:8083', offset=4, connectorIds=[], taskIds=[connector1-0], revokedConnectorIds=[], revokedTaskIds=[], delay=0} (org.apache.kafka.connect.runtime.distributed.WorkerCoordinator:200)
[2024-04-10 16:35:06,137] INFO Group coordinator localhost:1969 (id: 2147483647 rack: null) is unavailable or invalid due to cause: session timed out without receiving a heartbeat response. isDisconnected: false. Rediscovery will be attempted. (org.apache.kafka.connect.runtime.distributed.WorkerCoordinator:999)
[2024-04-10 16:35:06,137] INFO Requesting disconnect from last known coordinator localhost:1969 (id: 2147483647 rack: null) (org.apache.kafka.connect.runtime.distributed.WorkerCoordinator:1012)
[2024-04-10 16:35:06,138] DEBUG Sending FindCoordinator request to broker localhost:1969 (id: 0 rack: null) (org.apache.kafka.connect.runtime.distributed.WorkerCoordinator:904)
[2024-04-10 16:35:06,242] DEBUG Received FindCoordinator response ClientResponse(receivedTimeMs=1712747106110, latencyMs=29, disconnected=false, timedOut=false, requestHeader=RequestHeader(apiKey=FIND_COORDINATOR, apiVersion=4, clientId=mockClientId, correlationId=3, headerVersion=2), responseBody=HeartbeatResponseData(throttleTimeMs=0, errorCode=0)) (org.apache.kafka.connect.runtime.distributed.WorkerCoordinator:917)
[2024-04-10 16:35:06,245] ERROR Heartbeat thread failed due to unexpected error (org.apache.kafka.connect.runtime.distributed.WorkerCoordinator:1572)
java.lang.ClassCastException: org.apache.kafka.common.requests.HeartbeatResponse cannot be cast to org.apache.kafka.common.requests.FindCoordinatorResponse
	at org.apache.kafka.clients.consumer.internals.AbstractCoordinator$FindCoordinatorResponseHandler.onSuccess(AbstractCoordinator.java:919)
	at org.apache.kafka.clients.consumer.internals.AbstractCoordinator$FindCoordinatorResponseHandler.onSuccess(AbstractCoordinator.java:913)
	at org.apache.kafka.clients.consumer.internals.RequestFuture$1.onSuccess(RequestFuture.java:206)
	at org.apache.kafka.clients.consumer.internals.RequestFuture.fireSuccess(RequestFuture.java:169)
	at org.apache.kafka.clients.consumer.internals.RequestFuture.complete(RequestFuture.java:129)
	at org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient$RequestFutureCompletionHandler.fireCompletion(ConsumerNetworkClient.java:616)
	at org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient.firePendingCompletedRequests(ConsumerNetworkClient.java:428)
	at org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient.poll(ConsumerNetworkClient.java:313)
	at org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient.pollNoWakeup(ConsumerNetworkClient.java:322)
	at org.apache.kafka.clients.consumer.internals.AbstractCoordinator$HeartbeatThread.run(AbstractCoordinator.java:1500)
[2024-04-10 16:35:06,249] DEBUG Heartbeat thread has closed (org.apache.kafka.connect.runtime.distributed.WorkerCoordinator:1578)
[2024-04-10 16:35:21,119] INFO Metrics scheduler closed (org.apache.kafka.common.metrics.Metrics:694)
[2024-04-10 16:35:21,124] INFO Metrics reporters closed (org.apache.kafka.common.metrics.Metrics:704)
```

**Comment by vamossagar12** [comment]
@showuon , just checking did you get a chance to look at the updated test?

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/WorkerCoordinatorTest.java
Comment (showuon) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/WorkerCoordinatorTest.java]: Should we remove this line?

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/WorkerCoordinatorTest.java
Comment (showuon) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/WorkerCoordinatorTest.java]: Hmm.. I see. The reason is because we didn't make sure the heartbeat is sent after `time.sleep(sessionTimeoutMs - 1);`. Do you think we could make sure the heartbeat request or response is sent before next sleep? With that, we can make sure the session timeout won't happen because we did send out heartbeat every sessionTimeoutMs - 1.

Comment (showuon) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/WorkerCoordinatorTest.java]: What I want to say is, with your current change, this test will work like what you did before, which is not sending heartbeat response at all, and forcing session timeout directly. That works though, it's just not readable and doesn't make sense (i.e. not normal experience).

Comment (vamossagar12) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/WorkerCoordinatorTest.java]: yeah I got that. I added heartbeat request response till the rebalance timeout happens.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/WorkerCoordinatorTest.java
Comment (vamossagar12) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/WorkerCoordinatorTest.java]: For the purpose of this test, I create a new coordinator with rebalance timeout lesser than session timeout . As mentioned in the comment, it mayn't happen in the real world, but it makes testing a lot easier. Let me know what you think

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/WorkerCoordinatorTest.java
Comment (vamossagar12) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/WorkerCoordinatorTest.java]: This sleep value, though small, was the one that eventually worked after running the test continuously for 30 times. Sleeping for 10s still kept failing because sometimes the coordinator's session timeout would happen even though it doesn't look like it should.

**Comment by vamossagar12** [comment]
@showuon , I made some changes to the test so that we just heartbeat and not have to rely upon FindCoordinator requests. Let me know what yo think. Thanks!

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/WorkerCoordinatorTest.java
Comment (showuon) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/WorkerCoordinatorTest.java]: If you set `RebalanceTimeout < sessionTimeout`, why do we need the heartBeatRequest? You should be able to sleep (RebalanceTimeout + 1) directly to get the log. Ex:

```
try (LogCaptureAppender logCaptureAppender = LogCaptureAppender.createAndRegister(WorkerCoordinator.class)) {
            coordinator.ensureActiveGroup();
            coordinator.poll(0, () -> null);

             time.sleep(smallRebalanceTimeout + 1);

            // The heartbeat thread is running and keeps sending heartbeat requests.
            TestUtils.waitForCondition(() -> logCaptureAppender.getEvents().stream().anyMatch(e -> e.getLevel().equals("WARN")) &&
                    logCaptureAppender.getEvents().stream().anyMatch(e -> e.getMessage().startsWith("worker poll timeout has expired"), ...)
```

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/WorkerCoordinator.java
Comment (showuon) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/WorkerCoordinator.java]: @gharris1727 @mimaison , could you help review these log output since you're the connect experts? Thanks.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/WorkerCoordinatorTest.java
Comment (vamossagar12) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/WorkerCoordinatorTest.java]: I tried that, it still threw session timeout once in a while when I ran it for 30 times . So to be in the safer side, I added it

Comment (vamossagar12) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/WorkerCoordinatorTest.java]: I can try again to validate it

Comment (showuon) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/WorkerCoordinatorTest.java]: Yes, it should not throw session timeout anymore if you take my change above, because we only sleep once with `smallRebalanceTimeout + 1` ms, so the time will never exceed sessionTimeout. Does that make sense?

Comment (vamossagar12) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/WorkerCoordinatorTest.java]: @showuon , actually I typed in the wrong message yesterday. I ran the tests multiple times and it passes as expected. The reason I tried doing the other approach was that in the other [comment](https://github.com/apache/kafka/pull/15305#discussion_r1570226078) you had mentioned that it is not normal for consumers to not send heartbeats (other than readability). So to be closer to the real case, I had kept it.

Nonetheless I have reverted the code based on your suggestion and it works as well. Thanks!

**Review by showuon — APPROVED** [review]
LGTM! But I still need to wait until connect experts making sure this change makes sense. Thanks.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/WorkerCoordinator.java
Comment (vamossagar12) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/WorkerCoordinator.java]: Mickael/Greg could you please review the above message to see if it makes sense for connect? Thanks!

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/WorkerCoordinator.java
Comment (gharris1727) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/WorkerCoordinator.java:271]: Since we (as maintainers) don't have good insight into what commonly causes the herder tick thread to block and the poll timeout to fire, we recently added https://issues.apache.org/jira/browse/KAFKA-15563 to help users debug it themselves.

It would be nice to integrate with this system to have the heartbeat thread report what the herder tick thread was blocked on at the time that the poll timeout happened, as this would report stalling that isn't caused by REST requests.

The integration is tricky though, because the WorkerCoordinator is (and should be) unaware of the DistributedHerder. And currently I think the WorkerCoordinator hides these internal disconnects and reconnects inside of the poll method. Perhaps we can extend the WorkerRebalanceListener or have a new error listener to allow the herder to be informed about these errors.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/WorkerCoordinator.java
Comment (gharris1727) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/WorkerCoordinator.java]: I think this is decent advice when requests are small and can be distributed around the cluster, but as REST requests are rather infrequent, I think this is the minority of cases.

I think most often this timeout is going to be triggered by an excessively slow connector start, stop, or validation. In those cases, adding more workers does nothing but move the error to a different worker. I think we can keep the "adding more workers" comment, if we include another piece of advice for debugging excessively blocking tasks. If we don't have that other piece of advice, then advising users to add workers is misleading.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/WorkerCoordinator.java
Comment (vamossagar12) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/WorkerCoordinator.java:271]: Thanks Greg, I think that makes sense. I have extended the `WorkerRebalanceListener` to add another hook upon poll timeout expiry which the coordinator invokes. This way, it is able to access the tick thread's stage which was being executed at that point of time. I am just thinking, could there be race conditions where what we get in the tick thread stage might not always reflect the point where tick thread is blocked (like it moved on by the time we invoke this or becomes null), but I guess it should be ok because this can be treated as best effort? WDYT?

**Comment by vamossagar12** [comment]
Thanks @gharris1727 , I have made the changes as you suggested. I also modified the warning line that is printed based on the tick thread stage that is presented to the coordinator and also added a test to verify the same. LMKWYT.

File: connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/ConnectAssertions.java
Comment (vamossagar12) [inline connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/ConnectAssertions.java:79]: I tokk the liberty and edited this comment in this PR. Hope that's ok.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/WorkerCoordinator.java
Comment (gharris1727) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/WorkerCoordinator.java:271]: Yeah the tick thread stage is a best effort and can be misleading. However, if something has blocked long enough to cause a poll timeout, its likely enough to continue for the additional time it requires for the heartbeat thread to notice.

I think the ideal use-case I see is that this error pops up in a worker log 3 or more times before an operator has a chance to remediate it, and if the majority of logs blame connector-xyz, the operator can STOP that connector.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java
Comment (gharris1727) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java]: This method name does not make sense. The name would indicate that this method is used to inform the callee (the listener, this implementation) that the poll timeout happened. The actual behavior and signature is a getter for the current stage.

I don't think it makes sense for the listener to require the caller to keep a stage, so i think the implementation should change to follow the method name.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java
Comment (gharris1727) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: I would respect the "visible for testing" comment above, and leave this package-local.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/WorkerRebalanceListener.java
Comment (gharris1727) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/WorkerRebalanceListener.java]: This is an internal interface, unless this default method actually makes sense on it's own I wouldn't add it.

File: connect/runtime/src/test/java/org/apache/kafka/connect/integration/ConnectWorkerIntegrationTest.java
Comment (gharris1727) [inline connect/runtime/src/test/java/org/apache/kafka/connect/integration/ConnectWorkerIntegrationTest.java:886]: Could this assertion be added to an existing BlockingConnectorTest? The blocking plugins are inherently slow to use so we should avoid adding more instances of them.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/WorkerCoordinator.java
Comment (gharris1727) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/WorkerCoordinator.java]: From a user perspective, the class and method names are irrelevant, and bringing up irrelevant details in diagnostics can be misleading. "The last thing the worker was doing was: {} and may contribute to this timeout" is much more understandable and still gets the point across.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java
Comment (vamossagar12) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: This change is no longer valid. I was trying out some things to make the `WorkerCoordinator` test work and this came through because of that. Will remove.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java
Comment (vamossagar12) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java]: Yeah it seemed a bit weird to me as well but i thought that the log line for worker poll timeout needs to be added in `WorkerCoordinator` only. I have changed it to have it now being printed in the `WorkerRebalanceListener`. Let me know if it makes sense and that was something you had in mind.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/WorkerRebalanceListener.java
Comment (vamossagar12) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/WorkerRebalanceListener.java]: I added the default bit to avoid changing the tests. Turns out it should be ok to modify them so I removed the default bit.

File: connect/runtime/src/test/java/org/apache/kafka/connect/integration/ConnectWorkerIntegrationTest.java
Comment (vamossagar12) [reply connect/runtime/src/test/java/org/apache/kafka/connect/integration/ConnectWorkerIntegrationTest.java:886]: Yeah I wanted to add the test in `BlockingConnectorTest` itself but it would have meant a lot of changes in that class. That is because currently that test doesn't support setting worker level properties or changing the number of workers. Being able to change the worker level properties was the way I could get the poll timeout expiry. 
Moreover, the test I have added doesn't really block for the entire stop method but ends almost after the task shutdown graceful ms period ends because of the reset at the end of the test. Let me know if that makes sense.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/WorkerCoordinator.java
Comment (vamossagar12) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/WorkerCoordinator.java]: Got it, yeah I changed the message accordingly. I tried to keep the logline similar to the one in consumer's poll timeout but i think it's fine to deviate in this case.

**Comment by vamossagar12** [comment]
Thanks @gharris1727 for another round of review. I have made the changes. Please review when you have some time.

**Review by gharris1727 — APPROVED** [review]

**Comment by gharris1727** [comment]
Test failures appear unrelated, and the tests pass for me locally. LGTM, and thanks @vamossagar12 for the fix!

**Pull request merged**
